local/dependencies.go

```diff
@@ -25,37 +25,59 @@ import (
 	"golang.org/x/sync/errgroup"
 )
 
-func inDependencyOrder(ctx context.Context, project *types.Project, fn func(types.ServiceConfig) error) error {
-	eg, _ := errgroup.WithContext(ctx)
+func inDependencyOrder(ctx context.Context, project *types.Project, fn func(context.Context, types.ServiceConfig) error) error {
 	var (
 		scheduled []string
 		ready     []string
 	)
+	services := sortByDependency(project.Services)
+
 	results := make(chan string)
-	for len(ready) < len(project.Services) {
-		for _, service := range project.Services {
+	errors := make(chan error)
+	eg, ctx := errgroup.WithContext(ctx)
+	for len(ready) < len(services) {
+		for _, service := range services {
 			if contains(scheduled, service.Name) {
 				continue
 			}
 			if containsAll(ready, service.GetDependencies()) {
 				service := service
 				scheduled = append(scheduled, service.Name)
 				eg.Go(func() error {
-					err := fn(service)
+					err := fn(ctx, service)
 					if err != nil {
-						close(results)
+						errors <- err
```

Comment:
Not sure I understand why we need the `errors` chan

Reply:
other goroutine will try to close an already close channel, which results into panic

Reply:
for docker/c2 I uses a chan Pair(result, error) for this same needs.
I'd welcome any canonical golang reference about running things in parallel, collect results (which triggers more things to run, this is not just a parallel "map reduce") and stop on first error

Reply:
But wait, things are not running in parallel here right? Your code is equivalent to

```go
func inDependencyOrder(ctx context.Context, project *types.Project, fn func(context.Context, types.ServiceConfig) error) error {
	var (
		scheduled []string
		ready     []string
	)
	services := sortByDependency(project.Services)

	results := []string{}
	for len(ready) < len(services) {
		eg, _ := errgroup.WithContext(ctx)
		for _, service := range services {
			if contains(scheduled, service.Name) {
				continue
			}
			if containsAll(ready, service.GetDependencies()) {
				service := service
				scheduled = append(scheduled, service.Name)
				eg.Go(func() error {
					err := fn(ctx, service)
					if err != nil {
						return err
					}
					results = append(results, service.Name)
					return nil
				})
			}
		}

		if err := eg.Wait(); err != nil {
			return err
		}
	}

	return nil
}
```

Basically, each time you run a goroutine you just block while it goes up no? What am I missing here?

Reply:
Ok no, you need the `results` chan but you can move the `eg` inside the first `for` loop I think

Reply:
the idea is that first result _may_ allow to schedule another service to start.
example: with services A,B,C,D A depends on B, C depends on D
first iteration will create B and D in parallel
B starts first and sends results -> new iteration will trigger A to be created, while D might not (yet) be ready - especially considering support for `service_healthy`

note: this is mostly a direct transcription of docker-compose convergence logic here.

Reply:
Would be nice this loop actually run in parallel with the pull_policy one, so that some service can start and get healthy while other still pull... maybe a bit over-engineering for something that only will take seconds ;P

Reply:
Oookay understood, it's hairy :D

Reply:
Yes. The issue is the actual parrapel task list depends on the completed ones... This is where a dependency graph would help ;)